apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt

```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
     }
   }
 
+  /**
+   * A [BufferedSource] that writes to the given cache sink as it is read.
+   */
+  private class WriteToCacheSource(
+      private val originalBody: Source,
+      private val cacheSink: Sink,
+      private val cacheEditor: DiskLruCache.Editor,
+  ) : Source by originalBody {
+
+    private var hasClosedAndCommitted: Boolean = false
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      val buffer = Buffer()
+      val read = try {
+        originalBody.read(buffer, byteCount)
+      } catch (e: Exception) {
+        cacheEditor.abort()
+        throw e
+      }
+
+      if (read == -1L) {
+        // We've read fully, commit the cache edit
```

Comment:
Not 100% sure this is needed.

Reply:
The `DeferJvmTest` tests still pass for me without the `closeAndCommitCache()` can you add one that fails?

Reply:
Did we ever find the reason behind this?

Reply:
Sorry I thought your previous comment was about the test passing without committing because it wasn't checking the presence of data in the cache - but now I'm wondering if you meant something else 😅

Reply:
Let me try again

Reply:
Still the case. What I'm doing is I'm commenting out the `closeAndCommitCache()` on line 129. And then run tests:

```
./gradlew -p tests build
```

And everything is still ✅  leaving me to figure out what this is for? Or do I need to run other tests?

Reply:
Ah got it (thought you were commenting the inside of `closeAndCommitCache`).

This path can be reached when an incomplete response is fully read: the JSON reader pulls more bytes, but there aren't any, so `originalSource.read` returns -1. In that case this bit of code commits early, but:

1. it's actually not useful, because `close` will still be called after a parse exception occurs (which commits at that point)
2. not 100% sure if we actually want to commit in that case? On the one hand, this would be consistent with the current behavior (currently, the data is cached before it's parsed). On the other hand, if we know for sure that -1 here means an incomplete response, it may be nice to avoid caching that?

WDYT?

Reply:
Fell into this rabbit hole and turns out it's not too hard to implement. I gave it a try there: https://github.com/apollographql/apollo-kotlin/pull/4087. This is wildly untested. Feel free to include in this PR or I'll rebase and open a follow up PR later

Reply:
Yes this looks good! 🙏  Merging into this one and I'll add 2 tests for it.

Reply:
hmm 2 tests failed - I'll have a look before merging :P